src/Shared/NativeMethodsShared.cs

```diff
@@ -526,6 +547,57 @@ internal static string FixFrameworkPath(string path)
             return string.IsNullOrEmpty(m.Groups[3].Value) ? root + lastChar : Path.Combine(root, m.Groups[3].Value);
         }
 
+        /// <summary>
+        /// Set to true after processor architecture is determined.
+        /// When true, the values of s_processorArchitecture and
+        /// s_processorArchitectureNative are valid
+        /// </summary>
+        private static bool s_processorArchitectureSet = false;
+
+        /// <summary>
+        /// Architecture as far as the current process is concerned.
+        /// It's x86 in wow64 (native architecture is x64 in that case).
+        /// Otherwise it's the same as the native architecture.
+        /// </summary>
+        private static ProcessorArchitectures s_processorArchitecture;
+
+        /// <summary>
+        /// Actual architecture of the the system.
+        /// </summary>
+        private static ProcessorArchitectures s_processorArchitectureNative;
+
+        /// <summary>
+        /// Architecture getter
+        /// </summary>
+        internal static ProcessorArchitectures ProcessorArchitecture
+        {
+            get
+            {
+                if (!s_processorArchitectureSet)
```

Comment:
I thought about it, but decided it's not worth it. With a single property I don't think it would cause any harm if two threads try to run the code at the same time and update it to the same value. As it is, the first call is made way before any useful work is done. I think if we need more properties from the structure (number of cores, for example) and chances of this being called from multiple threads increase, we should do some locking.